### DISCUSSION TITLE
I think that each auth collection should have its own email templates

### DISCUSSION THREAD

**Discussion by dschissler** [pocketbase/pocketbase#1547] [Q&A]
I'm thinking that each auth collection should store the email templates (verification, password reset, and confirm email change) instead of it being centrally stored in mail settings.  The reason is that it makes it flexible to be able to have different types of user invitation designs that can utilize the built-in functionality, and without needing to worry too much about the ultimate design.

I mean that with each collection having its own email templates that a system can very easily be changed from closed and only via (closed and private) invitation, to open sign up, and also changing it in reverse (as the system changes).  It's also then possible to have different types of invitations that are from different collections.

With built-in email templates then there could be a closed invitation system that becomes obsolete.  Instead of needing to mutate that system into another system, that instead another invitation collection could be created.  It's flexible.  This could also accommodate different marketing efforts such that they don't need to relate or interfere with each other.

The way that this design pattern could work is that a record event `post update` hook from an invitation collection would then create the actual `users` user.

I think that also for this pattern to be the most effective that there would need to be a way for extra model data from the http request that doesn't correspond to any collection field to be available in the model event (instead of only registered table fields being set).  For example, there could be a required user field that needs to be set via the invitation form, but it's not a field in the invitation collection.

So with a typical system I'd create a user that isn't filled out and then I'd have a flag on the user such that it's valid until something is toggled.  That is ok, but it becomes very annoying to always be checking for user validity for everything from that point onwards.  However, since you've made Pocketbase be able to have unlimited auth collections then this can be much more easily controlled.  The limitation though is on the single email templates being used for the entire system.

As far as the mail settings UI, it could still be displayed as it is now when there is only a single auth collection, and then when there is more than one it could list those collections out as list with collection headers.  Also the email templates could be actually stored in each auth collection's meta data.  Then the mail settings for SMTP continue to be in the central mail settings table as they are now.

So the email templates per collection would be available in an auth collection's `Options` tab.

**Comment by dschissler** [comment]
Also the DB migration is very easy.  Just copy the current central mail templates over to each collection, and then delete it from the central settings.

**Comment by ganigeorgiev** [comment]
The benefit of having an individual auth email templates at the moment are very small and not worth investing dev effort in my opinion. I left the OAuth2 providers and email templates settings out of the auth collections intentionally to keep things as simple as possible.

This eventually may change in the future if we find a better and non-verbose way to structure it but other tasks are with higher priority right now. In general there are plans to improve and simplify the emails/notifications flow by adding "triggers" (or maybe even a workflow builder?) in https://github.com/pocketbase/pocketbase/issues/317 but I have to think a little more on it because it may not be necessary once we integrate the goja hooks in v0.13/v0.14.

**Reply by dschissler** [reply]
Will this "webhooks" be Javascript code that is store in the database and run on the server side?

Another possibility if it goes very general like this is that you could create a `templates` collection that is initially populated with several values / email templates.

I'll have to read more about these future plans.  Of course there are dramatic ways to make this more flexible and more complicated, so I'll have to get up to speed on what that is looking like.  As long as we're in agreement that the mail needs to be more flexible somehow then I'm sure h

**Reply by ganigeorgiev** [reply]
> Will this "webhooks" be Javascript code that is store in the database and run on the server side?

The triggers as defined in #317 will be primarily a Go code that will work with predefined placeholders (eg. the collection fields, the request/hook static data, etc.), but as mentioned in the above comment they may turn out redundant once we integrate goja with the app event hooks.

**Reply by dschissler** [reply]
The benefit that I can see in having the email templates per collection instead of central is that it could be possible to take advantage of the built-in Javascript API of `/request-verification` and other things for invitations.  It's completely possible to do it by making custom API's, but I think that this would allow a simple pattern that can almost entirely be done in tutorial documentation from the admin UI, and then with only a very simple custom event hook done with either Go or JS server-side to turn that verified invite into a new user record.

So with each collection having it's own email template for built-in functionality then it provides for more use cases that are very simple and don't require complex custom Framework usage.

Also in terms of these hooks, are the templates going to be embedded into the code, or are the templates still separate?  It seems strangely limited to have each collection have it's own auth capabilities, and also embeded scripting language support, but then to be hamstrung by a central email template.

I'm thinking that it could be nice to have `invitations` as an auth collection, and disabling oauth2, username and email logins.  So that it only functions as a mechanism for utilizing built-in email verification for private / marketing invites.  This is incredibly simple compared to creating custom API routes or overriding a few mailer events with a custom messages.

I'm thinking that there are so many dead simple patterns that could happen as a result of this.

But I'm still confused if you plan on removing the central email templates with this plan.
